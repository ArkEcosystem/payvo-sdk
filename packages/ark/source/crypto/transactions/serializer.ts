import { ByteBuffer } from "../crypto/buffer.js";

import { TransactionTypeGroup } from "../enums.js";
import { TransactionVersionError } from "../errors.js";
import { ISerializeOptions, ITransaction, ITransactionData } from "../interfaces/index.js";
import { configManager } from "../managers/config.js";
import { isSupportedTransactionVersion } from "../utils.js";
import { TransactionTypeFactory } from "./types/factory.js";

// Reference: https://github.com/ArkEcosystem/AIPs/blob/master/AIPS/aip-11.md
export class Serializer {
	public static getBytes(transaction: ITransactionData, options: ISerializeOptions = {}): Buffer {
		const version: number = transaction.version || 2;

		if (options.acceptLegacyVersion || options.disableVersionCheck || isSupportedTransactionVersion(version)) {
			return this.serialize(TransactionTypeFactory.create(transaction), options);
		} else {
			throw new TransactionVersionError(version);
		}
	}

	/**
	 * Serializes the given transaction according to AIP11.
	 */
	public static serialize(transaction: ITransaction, options: ISerializeOptions = {}): Buffer {
		const buffer: ByteBuffer = new ByteBuffer(512);

		this.serializeCommon(transaction.data, buffer);
		this.serializeVendorField(transaction, buffer);

		const serialized: ByteBuffer | undefined = transaction.serialize(options);

		if (!serialized) {
			throw new Error();
		}

		const typeBuffer: ByteBuffer = serialized.flip();
		buffer.append(typeBuffer.toBuffer());

		this.serializeSignatures(transaction.data, buffer, options);

		const flippedBuffer: Buffer = buffer.flip().toBuffer();
		transaction.serialized = flippedBuffer;

		return flippedBuffer;
	}

	private static serializeCommon(transaction: ITransactionData, buffer: ByteBuffer): void {
		transaction.version = transaction.version || 0x01;
		if (transaction.typeGroup === undefined) {
			transaction.typeGroup = TransactionTypeGroup.Core;
		}

		console.log(buffer.toBuffer().byteLength);

		buffer.writeByte(0xff);
		buffer.writeByte(transaction.version);
		buffer.writeByte(transaction.network || configManager.get("network.pubKeyHash"));
		buffer.writeUint32(transaction.typeGroup);
		buffer.writeUint16(transaction.type);

		if (transaction.nonce) {
			buffer.writeUint64(transaction.nonce.toString());
		}

		if (transaction.senderPublicKey) {
			buffer.append(transaction.senderPublicKey, "hex");
		}

		buffer.writeUint64(transaction.fee.toString());
	}

	private static serializeVendorField(transaction: ITransaction, buffer: ByteBuffer): void {
		if (transaction.hasVendorField()) {
			const { data }: ITransaction = transaction;

			if (data.vendorField) {
				const vf: Buffer = Buffer.from(data.vendorField, "utf8");
				buffer.writeByte(vf.length);
				buffer.append(vf);
			} else {
				buffer.writeByte(0x00);
			}
		} else {
			buffer.writeByte(0x00);
		}
	}

	private static serializeSignatures(
		transaction: ITransactionData,
		buffer: ByteBuffer,
		options: ISerializeOptions = {},
	): void {
		if (transaction.signature && !options.excludeSignature) {
			buffer.append(transaction.signature, "hex");
		}

		const secondSignature: string | undefined = transaction.secondSignature || transaction.signSignature;

		if (secondSignature && !options.excludeSecondSignature) {
			buffer.append(secondSignature, "hex");
		}

		if (transaction.signatures && !options.excludeMultiSignature) {
			buffer.append(transaction.signatures.join(""), "hex");
		}
	}
}
