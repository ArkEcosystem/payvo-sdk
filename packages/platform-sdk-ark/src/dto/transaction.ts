import { Contracts, DTO } from "@arkecosystem/platform-sdk";
import { DateTime } from "@arkecosystem/platform-sdk-intl";
import { BigNumber } from "@arkecosystem/platform-sdk-support";

export class TransactionData extends DTO.AbstractTransactionData implements Contracts.TransactionData {
	public id(): string {
		return this.data.id;
	}

	public blockId(): string | undefined {
		return this.data.blockId;
	}

	public timestamp(): DateTime | undefined {
		return DateTime.fromUnix(this.data.timestamp.unix);
	}

	public confirmations(): BigNumber {
		return BigNumber.make(this.data.confirmations);
	}

	public sender(): string {
		return this.data.sender;
	}

	public recipient(): string {
		return this.data.recipient;
	}

	public recipients(): Contracts.MultiPaymentRecipient[] {
		if (!this.isMultiPayment()) {
			return [];
		}

		return this.data.asset.payments.map((payment: { recipientId: string; amount: string }) => ({
			address: payment.recipientId,
			amount: BigNumber.make(payment.amount),
		}));
	}

	public amount(): BigNumber {
		if (this.isMultiPayment()) {
			return this.data.asset.payments.reduce(
				(sum: BigNumber, { amount }: { amount: string }) => sum.plus(amount),
				BigNumber.ZERO,
			);
		}

		return BigNumber.make(this.data.amount);
	}

	public fee(): BigNumber {
		return BigNumber.make(this.data.fee);
	}

	public memo(): string | undefined {
		return this.data.vendorField;
	}

	public asset(): Record<string, unknown> {
		return this.data.asset;
	}

	public isConfirmed(): boolean {
		return this.confirmations().isGreaterThanOrEqualTo(51);
	}

	public isSent(): boolean {
		return [this.getMeta("address"), this.getMeta("publicKey")].includes(this.sender());
	}

	public isReceived(): boolean {
		return [this.getMeta("address"), this.getMeta("publicKey")].includes(this.recipient());
	}

	public isTransfer(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 0;
	}

	public isSecondSignature(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 1;
	}

	public isDelegateRegistration(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 2;
	}

	public isVote(): boolean {
		const isVote = this.data.typeGroup === 1 && this.data.type === 3;

		if (!isVote) {
			return false;
		}

		return (this.asset().votes as string[])[0].startsWith("+");
	}

	public isUnvote(): boolean {
		const isVote = this.data.typeGroup === 1 && this.data.type === 3;

		if (!isVote) {
			return false;
		}

		return (this.asset().votes as string[])[0].startsWith("-");
	}

	public isMultiSignature(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 4;
	}

	public isIpfs(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 5;
	}

	public isMultiPayment(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 6;
	}

	public isDelegateResignation(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 7;
	}

	public isHtlcLock(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 8;
	}

	public isHtlcClaim(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 9;
	}

	public isHtlcRefund(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 10;
	}

	public isBusinessRegistration(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 0;
	}

	public isBusinessResignation(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 1;
	}

	public isBusinessUpdate(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 2;
	}

	public isBridgechainRegistration(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 3;
	}

	public isBridgechainResignation(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 4;
	}

	public isBridgechainUpdate(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 5;
	}

	public isEntityRegistration(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 6 && this.data.asset.action === 0;
	}

	public isEntityResignation(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 6 && this.data.asset.action === 1;
	}

	public isEntityUpdate(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 6 && this.data.asset.action === 2;
	}
}
