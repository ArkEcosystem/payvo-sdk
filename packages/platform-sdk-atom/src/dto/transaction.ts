import { Contracts, DTO, Exceptions } from "@arkecosystem/platform-sdk";
import { DateTime } from "@arkecosystem/platform-sdk-intl";
import { BigNumber } from "@arkecosystem/platform-sdk-support";

export class TransactionData extends DTO.AbstractTransactionData implements Contracts.TransactionData {
	public id(): string {
		return this.data.txhash;
	}

	public blockId(): string | undefined {
		return undefined;
	}

	public timestamp(): DateTime | undefined {
		return DateTime.make(this.data.timestamp);
	}

	public confirmations(): BigNumber {
		return BigNumber.ZERO;
	}

	public sender(): string {
		const event = this.data.events.find((event) => event.type === "message");
		const attribute = event.attributes.find((attribute) => attribute.key === "sender");

		return attribute.value;
	}

	public recipient(): string {
		const event = this.data.events.find((event) => event.type === "transfer");
		const attribute = event.attributes.find((attribute) => attribute.key === "recipient");

		return attribute.value;
	}

	public recipients(): Contracts.MultiPaymentRecipient[] {
		return [];
	}

	// @ts-ignore
	public amount(): BigNumber {
		const event = this.data.events.find((event) => event.type === "transfer");
		const attribute = event.attributes.find((attribute) => attribute.key === "amount");

		return BigNumber.make(attribute.value.replace(/\D/g, ""));
	}

	public fee(): BigNumber {
		return BigNumber.make(this.data.gas_used);
	}

	public memo(): string | undefined {
		return this.data.tx.value.memo;
	}

	public asset(): Record<string, unknown> {
		return {};
	}

	public isConfirmed(): boolean {
		return false;
	}

	public isSent(): boolean {
		return false;
	}

	public isReceived(): boolean {
		return false;
	}

	public isTransfer(): boolean {
		return false;
	}

	public isSecondSignature(): boolean {
		return false;
	}

	public isDelegateRegistration(): boolean {
		return false;
	}

	public isVote(): boolean {
		return false;
	}

	public isUnvote(): boolean {
		return false;
	}

	public isMultiSignature(): boolean {
		return false;
	}

	public isIpfs(): boolean {
		return false;
	}

	public isMultiPayment(): boolean {
		return false;
	}

	public isDelegateResignation(): boolean {
		return false;
	}

	public isHtlcLock(): boolean {
		return false;
	}

	public isHtlcClaim(): boolean {
		return false;
	}

	public isHtlcRefund(): boolean {
		return false;
	}

	public isEntityRegistration(): boolean {
		return false;
	}

	public isEntityResignation(): boolean {
		return false;
	}

	public isEntityUpdate(): boolean {
		return false;
	}

	public isBusinessEntityRegistration(): boolean {
		return false;
	}

	public isBusinessEntityResignation(): boolean {
		return false;
	}

	public isBusinessEntityUpdate(): boolean {
		return false;
	}

	public isDeveloperEntityRegistration(): boolean {
		return false;
	}

	public isDeveloperEntityResignation(): boolean {
		return false;
	}

	public isDeveloperEntityUpdate(): boolean {
		return false;
	}

	public isCorePluginEntityRegistration(): boolean {
		return false;
	}

	public isCorePluginEntityResignation(): boolean {
		return false;
	}

	public isCorePluginEntityUpdate(): boolean {
		return false;
	}

	public isDesktopPluginEntityRegistration(): boolean {
		return false;
	}

	public isDesktopPluginEntityResignation(): boolean {
		return false;
	}

	public isDesktopPluginEntityUpdate(): boolean {
		return false;
	}

	public isDelegateEntityRegistration(): boolean {
		return false;
	}

	public isDelegateEntityResignation(): boolean {
		return false;
	}

	public isDelegateEntityUpdate(): boolean {
		return false;
	}

	public isLegacyBusinessRegistration(): boolean {
		return false;
	}

	public isLegacyBusinessResignation(): boolean {
		return false;
	}

	public isLegacyBusinessUpdate(): boolean {
		return false;
	}

	public isLegacyBridgechainRegistration(): boolean {
		return false;
	}

	public isLegacyBridgechainResignation(): boolean {
		return false;
	}

	public isLegacyBridgechainUpdate(): boolean {
		return false;
	}
}
