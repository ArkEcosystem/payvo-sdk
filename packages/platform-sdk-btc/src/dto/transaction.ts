import { Contracts, DTO, Exceptions } from "@arkecosystem/platform-sdk";
import { DateTime } from "@arkecosystem/platform-sdk-intl";
import { BigNumber } from "@arkecosystem/platform-sdk-support";

export class TransactionData extends DTO.AbstractTransactionData implements Contracts.TransactionData {
	public id(): string {
		return this.data.txid;
	}

	public blockId(): string | undefined {
		return undefined;
	}

	public timestamp(): DateTime | undefined {
		return DateTime.make(this.data.blockTime);
	}

	public confirmations(): BigNumber {
		return BigNumber.make(this.data.confirmations);
	}

	public sender(): string {
		throw new Exceptions.NotImplemented(this.constructor.name, "sender");
	}

	public recipient(): string {
		throw new Exceptions.NotImplemented(this.constructor.name, "recipient");
	}

	public recipients(): Contracts.MultiPaymentRecipient[] {
		return [];
	}

	public amount(): BigNumber {
		return BigNumber.make(this.data.value);
	}

	public fee(): BigNumber {
		return BigNumber.make(this.data.fee);
	}

	public memo(): string | undefined {
		return undefined;
	}

	public asset(): Record<string, unknown> {
		return {};
	}

	public isConfirmed(): boolean {
		return false;
	}

	public isSent(): boolean {
		return false;
	}

	public isReceived(): boolean {
		return false;
	}

	public isTransfer(): boolean {
		return false;
	}

	public isSecondSignature(): boolean {
		return false;
	}

	public isDelegateRegistration(): boolean {
		return false;
	}

	public isVote(): boolean {
		return false;
	}

	public isUnvote(): boolean {
		return false;
	}

	public isMultiSignature(): boolean {
		return false;
	}

	public isIpfs(): boolean {
		return false;
	}

	public isMultiPayment(): boolean {
		return false;
	}

	public isDelegateResignation(): boolean {
		return false;
	}

	public isHtlcLock(): boolean {
		return false;
	}

	public isHtlcClaim(): boolean {
		return false;
	}

	public isHtlcRefund(): boolean {
		return false;
	}

	public isEntityRegistration(): boolean {
		return false;
	}

	public isEntityResignation(): boolean {
		return false;
	}

	public isEntityUpdate(): boolean {
		return false;
	}

	public isBusinessEntityRegistration(): boolean {
		return false;
	}

	public isBusinessEntityResignation(): boolean {
		return false;
	}

	public isBusinessEntityUpdate(): boolean {
		return false;
	}

	public isDeveloperEntityRegistration(): boolean {
		return false;
	}

	public isDeveloperEntityResignation(): boolean {
		return false;
	}

	public isDeveloperEntityUpdate(): boolean {
		return false;
	}

	public isCorePluginEntityRegistration(): boolean {
		return false;
	}

	public isCorePluginEntityResignation(): boolean {
		return false;
	}

	public isCorePluginEntityUpdate(): boolean {
		return false;
	}

	public isDesktopPluginEntityRegistration(): boolean {
		return false;
	}

	public isDesktopPluginEntityResignation(): boolean {
		return false;
	}

	public isDesktopPluginEntityUpdate(): boolean {
		return false;
	}

	public isDelegateEntityRegistration(): boolean {
		return false;
	}

	public isDelegateEntityResignation(): boolean {
		return false;
	}

	public isDelegateEntityUpdate(): boolean {
		return false;
	}

	public isLegacyBusinessRegistration(): boolean {
		return false;
	}

	public isLegacyBusinessResignation(): boolean {
		return false;
	}

	public isLegacyBusinessUpdate(): boolean {
		return false;
	}

	public isLegacyBridgechainRegistration(): boolean {
		return false;
	}

	public isLegacyBridgechainResignation(): boolean {
		return false;
	}

	public isLegacyBridgechainUpdate(): boolean {
		return false;
	}
}
