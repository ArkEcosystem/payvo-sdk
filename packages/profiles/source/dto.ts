export * from "./signed-transaction.dto.js";
export * from "./transaction.collection";
export * from "./transaction.dto.js";
export * from "./transaction.mapper";
