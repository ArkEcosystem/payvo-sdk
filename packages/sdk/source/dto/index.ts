export * from "./confirmed-transaction";
export * from "./signed-transaction";
export * from "./signed-transaction.contract";
export * from "./unspent-transaction";
export * from "./wallet";
