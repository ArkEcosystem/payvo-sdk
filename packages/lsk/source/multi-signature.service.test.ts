
import { DateTime } from "@payvo/sdk-intl";
import { IoC, Services, Signatories, Test } from "@payvo/sdk";
import nock from "nock";

import { identity } from "../test/fixtures/identity.js";
import { createService, requireModule } from "../test/mocking.js";
import { AddressService } from "./address.service.js";
import { ClientService } from "./client.service.js";
import { FeeService } from "./fee.service.js";
import { KeyPairService } from "./key-pair.service.js";
import { LedgerService } from "./ledger.service.js";
import { PublicKeyService } from "./public-key.service.js";
import { TransactionService } from "./transaction.service.js";
import { MultiSignatureService } from "./multi-signature.service.js";
import { TransactionSerializer } from "./transaction.serializer";
import { BindingType } from "./coin.contract.js";
import { AssetSerializer } from "./asset.serializer";
import { SignedTransactionData } from "./signed-transaction.dto.js";
import { ConfirmedTransactionData } from "./confirmed-transaction.dto.js";
import { WalletData } from "./wallet.dto.js";

let subject: TransactionService;
let musig: MultiSignatureService;

beforeAll(async () => {
    nock.disableNetConnect();

    subject = await createService(TransactionService, "lsk.testnet", (container) => {
        container.constant(IoC.BindingType.Container, container);
        container.singleton(IoC.BindingType.AddressService, AddressService);
        container.singleton(IoC.BindingType.ClientService, ClientService);
        container.constant(IoC.BindingType.DataTransferObjects, {
            SignedTransactionData,
            ConfirmedTransactionData,
            WalletData,
        });
        container.singleton(IoC.BindingType.DataTransferObjectService, Services.AbstractDataTransferObjectService);
        container.singleton(IoC.BindingType.FeeService, FeeService);
        container.singleton(IoC.BindingType.KeyPairService, KeyPairService);
        container.constant(IoC.BindingType.LedgerTransportFactory, async () => { });
        container.singleton(IoC.BindingType.LedgerService, LedgerService);
        container.singleton(IoC.BindingType.PublicKeyService, PublicKeyService);
        container.singleton(IoC.BindingType.MultiSignatureService, MultiSignatureService);
        container.singleton(BindingType.AssetSerializer, AssetSerializer);
        container.singleton(BindingType.TransactionSerializer, TransactionSerializer);
    });

    musig = createService(MultiSignatureService, "lsk.testnet", (container) => {
        container.constant(IoC.BindingType.Container, container);
        container.singleton(IoC.BindingType.AddressService, AddressService);
        container.singleton(IoC.BindingType.ClientService, ClientService);
        container.constant(IoC.BindingType.DataTransferObjects, {
            SignedTransactionData,
            ConfirmedTransactionData,
            WalletData,
        });
        container.singleton(IoC.BindingType.DataTransferObjectService, Services.AbstractDataTransferObjectService);
        container.singleton(IoC.BindingType.FeeService, FeeService);
        container.singleton(IoC.BindingType.KeyPairService, KeyPairService);
        container.constant(IoC.BindingType.LedgerTransportFactory, async () => { });
        container.singleton(IoC.BindingType.LedgerService, LedgerService);
        container.singleton(IoC.BindingType.PublicKeyService, PublicKeyService);
        container.singleton(BindingType.AssetSerializer, AssetSerializer);
        container.singleton(BindingType.TransactionSerializer, TransactionSerializer);
    });

    jest.spyOn(DateTime, "make").mockReturnValue(DateTime.make("2021-01-01 12:00:00"));
});

describe("MultiSignatureService", () => {
    const wallet1 = {
        signingKey: "foil broccoli rare pony man umbrella visual cram wing rotate fall never",
        address: "lskp4agpmjwgw549xdrhgdt6dfwqrpvohgbkhyt8p",
        publicKey: "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    };

    const wallet2 = {
        signingKey: "penalty name learn right reason inherit peace mango guitar heart nature love",
        address: "lskn2de9mo9z3g9jvbpj4yjn84vrvjzcn5c5mon7a",
        publicKey: "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
    };

    beforeAll(async () => {
        nock(/.+/)
            .get("/api/v2/accounts")
            .query({ address: "lskp4agpmjwgw549xdrhgdt6dfwqrpvohgbkhyt8p" })
            .reply(200, requireModule(`../test/fixtures/musig/lskp4agpmjwgw549xdrhgdt6dfwqrpvohgbkhyt8p.json`))
            .get("/api/v2/accounts")
            .query({ publicKey: "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed" })
            .reply(200, requireModule(`../test/fixtures/musig/lskp4agpmjwgw549xdrhgdt6dfwqrpvohgbkhyt8p.json`))
            .persist();
    });

    it("should add signature", async () => {
        const transaction1 = await subject.transfer({
            fee: 10,
            signatory: new Signatories.Signatory(
                new Signatories.MnemonicSignatory({
                    signingKey: wallet1.signingKey,
                    address: wallet1.address,
                    publicKey: wallet1.publicKey,
                    privateKey: identity.privateKey,
                }),
            ),
            data: {
                amount: 1,
                to: wallet1.address,
            },
        });

        assert.is(transaction1 instanceof SignedTransactionData);
        assert.is(transaction1).toMatchSnapshot();

        assert.is(musig.isMultiSignatureReady(transaction1), false);
        assert.is(musig.needsSignatures(transaction1), true);
        assert.is(musig.needsAllSignatures(transaction1), true);
        assert.is(musig.remainingSignatureCount(transaction1), 1);
        assert.is(musig.needsWalletSignature(transaction1, wallet1.publicKey), false);
        assert.is(musig.needsWalletSignature(transaction1, wallet2.publicKey), true);

        const transaction2 = await musig.addSignature(
            transaction1.data(),
            new Signatories.Signatory(
                new Signatories.MnemonicSignatory({
                    signingKey: wallet2.signingKey,
                    address: wallet2.address,
                    publicKey: wallet2.publicKey,
                    privateKey: identity.privateKey,
                }),
            ),
        );

        assert.is(transaction2 instanceof SignedTransactionData);
        assert.is(transaction2).toMatchSnapshot();

        assert.is(musig.isMultiSignatureReady(transaction2), true);
        assert.is(musig.needsSignatures(transaction2), false);
        assert.is(musig.needsAllSignatures(transaction2), false);
        assert.is(musig.remainingSignatureCount(transaction2), 0);
        assert.is(musig.needsWalletSignature(transaction2, wallet1.publicKey), false);
        assert.is(musig.needsWalletSignature(transaction2, wallet2.publicKey), false);
    });

    describe("#broadcast", () => {
        let transaction;

        test.before.each(async () => {
            transaction = await musig.addSignature(
                (
                    await subject.transfer({
                        fee: 10,
                        signatory: new Signatories.Signatory(
                            new Signatories.MnemonicSignatory({
                                signingKey: wallet1.signingKey,
                                address: wallet1.address,
                                publicKey: wallet1.publicKey,
                                privateKey: identity.privateKey,
                            }),
                        ),
                        data: {
                            amount: 1,
                            to: wallet1.address,
                        },
                    })
                ).data(),
                new Signatories.Signatory(
                    new Signatories.MnemonicSignatory({
                        signingKey: wallet2.signingKey,
                        address: wallet2.address,
                        publicKey: wallet2.publicKey,
                        privateKey: identity.privateKey,
                    }),
                ),
            );
        });

        it("should broadcast a transaction", async () => {
            nock(/.+/)
                .post("/", (body) => body.method === "store")
                .reply(200, {
                    result: { id: transaction.id() },
                });

            await assert.is(musig.broadcast(transaction.data())).resolves.toEqual({
                accepted: [transaction.id()],
                errors: {},
                rejected: [],
            });
        });

        it("should handle error", async () => {
            nock(/.+/)
                .post("/", (body) => body.method === "store")
                .reply(400, {
                    message: "Unable to broadcast transaction.",
                });

            await assert.is(musig.broadcast(transaction.data())).resolves.toEqual({
                accepted: [],
                errors: {
                    [transaction.id()]: "Unable to broadcast transaction.",
                },
                rejected: [transaction.id()],
            });
        });
    });

    test("#needsFinalSignature", async () => {
        assert.is(
            musig.needsFinalSignature(
                await subject.transfer({
                    fee: 10,
                    signatory: new Signatories.Signatory(
                        new Signatories.MnemonicSignatory({
                            signingKey: wallet1.signingKey,
                            address: wallet1.address,
                            publicKey: wallet1.publicKey,
                            privateKey: identity.privateKey,
                        }),
                    ),
                    data: {
                        amount: 1,
                        to: wallet1.address,
                    },
                }),
            ),
        , true);
    });

    test("#allWithPendingState", async () => {
        nock(/.+/)
            .post("/", {
                jsonrpc: "2.0",
                id: /.+/,
                method: "list",
                params: {
                    publicKey: identity.publicKey,
                    state: "pending",
                },
            })
            .reply(200, {
                result: [
                    { data: {}, multiSignature: {} },
                    { data: {}, multiSignature: {} },
                ],
            });

        await assert.is(musig.allWithPendingState(identity.publicKey)).resolves.toHaveLength(2);
    });

    test("#allWithReadyState", async () => {
        nock(/.+/)
            .post("/", {
                jsonrpc: "2.0",
                id: /.+/,
                method: "list",
                params: {
                    publicKey: identity.publicKey,
                    state: "ready",
                },
            })
            .reply(200, {
                result: [
                    { data: {}, multiSignature: {} },
                    { data: {}, multiSignature: {} },
                ],
            });

        await assert.is(musig.allWithReadyState(identity.publicKey)).resolves.toHaveLength(2);
    });

    test("#findById", async () => {
        nock(/.+/)
            .post("/", {
                jsonrpc: "2.0",
                id: /.+/,
                method: "show",
                params: {
                    id: "384b0438-36c0-4437-a35b-a8135cbba17d",
                },
            })
            .reply(200, { result: { data: {}, multiSignature: {} } });

        await assert.is(musig.findById("384b0438-36c0-4437-a35b-a8135cbba17d")).resolves.toEqual({
            multiSignature: {},
        });
    });

    test("#forgetById", async () => {
        const deleteNock = nock(/.+/)
            .post("/", {
                jsonrpc: "2.0",
                id: /.+/,
                method: "delete",
                params: {
                    id: "384b0438-36c0-4437-a35b-a8135cbba17d",
                },
            })
            .reply(200, {});

        assert.is(deleteNock.isDone(), false);

        await musig.forgetById("384b0438-36c0-4437-a35b-a8135cbba17d");

        assert.is(deleteNock.isDone(), true);
    });
});
