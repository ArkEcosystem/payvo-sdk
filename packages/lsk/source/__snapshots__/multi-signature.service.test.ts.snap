// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSignatureService should add signature 1`] = `
SignedTransactionData {
  "bigNumberService": BigNumberService {
    "configRepository": ConfigRepository {},
  },
  "broadcastData": Object {
    "asset": Object {
      "amount": "100000000",
      "data": "",
      "recipientAddress": "lskp4agpmjwgw549xdrhgdt6dfwqrpvohgbkhyt8p",
    },
    "assetID": 0,
    "fee": "1000000000",
    "id": "7395fb8f6d92a80afc28047e07ce3be4cd2beb4be11d36584961b7c6c3ddd664",
    "moduleID": 2,
    "multiSignature": Object {
      "mandatoryKeys": Array [
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "nonce": "8",
    "senderPublicKey": "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    "signatures": Array [
      "",
      "9ccc8d65bfc3b777217ac8670b2e511f854c160e6a00b8e9a0869016657b97fbe115811e6229a4e8c8e46990d6e2603f796d108837c5b09d13f6ed6dad08ab0c",
    ],
    "timestamp": "2021-01-01T12:00:00.000Z",
  },
  "decimals": 8,
  "identifier": "7395fb8f6d92a80afc28047e07ce3be4cd2beb4be11d36584961b7c6c3ddd664",
  "signedData": Object {
    "asset": Object {
      "amount": "100000000",
      "data": "",
      "recipientAddress": "lskp4agpmjwgw549xdrhgdt6dfwqrpvohgbkhyt8p",
    },
    "assetID": 0,
    "fee": "1000000000",
    "id": "7395fb8f6d92a80afc28047e07ce3be4cd2beb4be11d36584961b7c6c3ddd664",
    "moduleID": 2,
    "multiSignature": Object {
      "mandatoryKeys": Array [
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "nonce": "8",
    "senderPublicKey": "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    "signatures": Array [
      "",
      "9ccc8d65bfc3b777217ac8670b2e511f854c160e6a00b8e9a0869016657b97fbe115811e6229a4e8c8e46990d6e2603f796d108837c5b09d13f6ed6dad08ab0c",
    ],
    "timestamp": "2021-01-01T12:00:00.000Z",
  },
}
`;

exports[`MultiSignatureService should add signature 2`] = `
SignedTransactionData {
  "bigNumberService": BigNumberService {
    "configRepository": ConfigRepository {},
  },
  "broadcastData": Object {
    "asset": Object {
      "amount": "100000000",
      "data": "",
      "recipientAddress": "lskp4agpmjwgw549xdrhgdt6dfwqrpvohgbkhyt8p",
    },
    "assetID": 0,
    "fee": "1000000000",
    "id": "9b251f6caf03725dcf6514dae3b147953fec70caff9db332953a1a02573e729d",
    "moduleID": 2,
    "multiSignature": Object {
      "mandatoryKeys": Array [
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "nonce": "8",
    "senderPublicKey": "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    "signatures": Array [
      "3760f52b99f5c78b4cd3f46374baef1774279c876d9f23abef50b60670983916c3d69a33f8dd51acb7ef9ee30c6704e0d31bfbd9a42b5f17ba1d6ac6d4811505",
      "9ccc8d65bfc3b777217ac8670b2e511f854c160e6a00b8e9a0869016657b97fbe115811e6229a4e8c8e46990d6e2603f796d108837c5b09d13f6ed6dad08ab0c",
    ],
    "timestamp": "2021-01-01T12:00:00.000Z",
  },
  "decimals": 8,
  "identifier": "9b251f6caf03725dcf6514dae3b147953fec70caff9db332953a1a02573e729d",
  "signedData": Object {
    "asset": Object {
      "amount": "100000000",
      "data": "",
      "recipientAddress": "lskp4agpmjwgw549xdrhgdt6dfwqrpvohgbkhyt8p",
    },
    "assetID": 0,
    "fee": "1000000000",
    "id": "9b251f6caf03725dcf6514dae3b147953fec70caff9db332953a1a02573e729d",
    "moduleID": 2,
    "multiSignature": Object {
      "mandatoryKeys": Array [
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "nonce": "8",
    "senderPublicKey": "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    "signatures": Array [
      "3760f52b99f5c78b4cd3f46374baef1774279c876d9f23abef50b60670983916c3d69a33f8dd51acb7ef9ee30c6704e0d31bfbd9a42b5f17ba1d6ac6d4811505",
      "9ccc8d65bfc3b777217ac8670b2e511f854c160e6a00b8e9a0869016657b97fbe115811e6229a4e8c8e46990d6e2603f796d108837c5b09d13f6ed6dad08ab0c",
    ],
    "timestamp": "2021-01-01T12:00:00.000Z",
  },
}
`;
